myia/ir/node.py

```diff
@@ -0,0 +1,183 @@
+from myia.utils import Named
+from myia.utils.info import DebugInfo
+
+FN = Named("$fn")
+SEQ = Named("$seq")
+
+
+class Graph:
+    def __init__(self, parent=None):
+        self.parent = parent
+        self.parameters = []
+        self.return_ = None
+        self.flags = {}
+        self.varargs = False
+        self.kwargs = False
+        self.defaults = {}
+        self.kwonly = 0
+        self.debug = DebugInfo(obj=self)
+
+    @property
+    def output(self):
+        if not self.return_ or 0 not in self.return_.edges:
+            raise ValueError("Graph has no output")
+        return self.return_.edges[0].node
+
+    @output.setter
+    def output(self, value):
+        if self.return_:
+            self.return_.edges[0].node = value
+        else:
+            self.return_ = self.apply("return_", value)
+        self.return_.abstract = value.abstract
+        # XXX: add typing for the "return_" primitive (or maybe not)
+
+    def set_flags(self, **flags):
+        self.flags.update(flags)
+
+    def add_parameter(self, name):
+        p = Parameter(self, name)
+        self.parameters.append(p)
+        return p
+
+    def constant(self, obj):
+        return Constant(obj)
+
+    def apply(self, fn, *inputs):
+        edges = [
+            Edge(p, i if isinstance(i, Node) else self.constant(i))
+            for p, i in enumerate(inputs)
+        ]
+        edges.append(
+            Edge(FN, fn if isinstance(fn, Node) else self.constant(fn))
+        )
+
+        return Apply(self, *edges)
+
+    def replace(self, mapping, mapping_seq={}):
+        todo = [self.return_]
+        seen = set()
+        while todo:
+            node = todo.pop()
+            if node in seen:
+                continue
+            seen.add(node)
+            for edge in node.edges.values():
+                if edge.label is SEQ and edge.node in mapping_seq:
+                    edge.node = mapping_seq[edge.node]
+                elif edge.node in mapping:
+                    edge.node = mapping[edge.node]
+                if edge.node and edge.node.is_apply():
+                    todo.append(edge.node)
```

Comment:
I'd like some documentation about what this function does and why it exists.

It might also be a good idea to add an `iter_edges` method that iterates over edges and use it to implement `replace`. Pretty sure it'd be useful in other situations.

Reply:
I did add some documentation in the latest commit, tell me if it's good enough.

As for iter_edges, I'm not sure.  I'm not against the idea, but I don't see the advantages over plain for loop.

Reply:
Why are there two mapping dictionaries? That's the main interrogation I have right now.

I meant an iter_edges method on Graph to perform a bfs of all of the graph's edges. I suppose we can refactor when we need to, but I suspect we will need to do it at some point.

Reply:
Because sometimes you want to replace a node with another one, but then sequence chain should flow differently.

So mapping_seq allows to override (or define) replacements for the sequence chain specifically.

Reply:
Okay, I think I see what you mean. Replacing node 1 by node 2 in the data flow might not mean you want to perform the same replacement in the sequence chain.